Don't return null for value type after DoesNothing

blairconrad opened the pull request:
Fixes #806

adamralph commented on src/FakeItEasy/Configuration/RuleBuilder.cs:
```diff
@@ -63,7 +63,6 @@ public IVoidConfiguration WhenArgumentsMatch(Func<ArgumentCollection, bool> argu
 
         public virtual IAfterCallSpecifiedConfiguration DoesNothing()
         {
-            this.RuleBeingBuilt.Applicator = x => { };
```
It seems so simple, yet it worries me at the same time for some reason I can't quite put my finger on.

Is there any undesired side effect here? I won't pretend to know what `this.RuleBeingBuilt.Applicator` is or does, but is there any scenario where this will change behaviour and break some existing tests? I.e. is it currently possible to chain `DoesNothing` onto some other configuration calls, which effectively clears some previous config? (Please tell me if I'm talking complete nonsense!)

blairconrad replied:
You're not talking nonsense at all. It is a little worrisome, isn't it?

First, I gotta say that the new implementation is a lot closer to "does nothing" than before. If only I could get rid of the pesky `return`, but we need something to hang the `Repeated` off.

Okay. I'll be serious now.

Your concerns are completely reasonable. 
The `Applicator` is... the thing that does stuff. Most of the time, it's used to set a return value. Otherwise, it's the thing that `Throws` the exception.
I think that's about it, really.

And every new rule is created with an applicator that returns the default value (null or value instance), so I think that aspect of the behaviour is correct.

Although you've caused me to realize something. A while back @thomaslevesque fixed #534 to make sure we couldn't save away a rule and then add multiple behaviours to it. The change as-is will break that for `DoesNothing`. Even though there's no "normal" way to make this happen using the fluent API, my change can't stand. I'll add an extra spec and amend the fix.

Thanks!

thomaslevesque replied:
There's another reason why this change doesn't really work : it no longer causes the call to do nothing if the call was doing something... With your current fix, if you do something like this:

```
var fake = A.Fake<SomeClass>();
A.CallTo(fake).CallsBaseMethod();
A.CallTo(() => fake.SomeMethod()).DoesNothing();
```
`DoesNothing` will _not_ prevent the call to `base.SomeMethod()`.

I think `DoesNothing` should reset the applicator to its initial value (the one set in the `BuildableCallRule` constructor).

blairconrad replied:
@thomaslevesque, in this example, `DoesNothing` will prevent the call to `base.SomeMethod()`. It works because the repeated `CallTo` creates a new rule on which the applicator lives. The `DoesNothing` will leave the default applicator on the rule, and then it's used when the call is made.

Only if we reused the result of the first `A.CallTo` would the `DoesNothing` not override the `CallsBaseMethod`, but this is the same situation that I described above, where I break #534.

thomaslevesque replied:
> @thomaslevesque, in this example, DoesNothing will prevent the call to base.SomeMethod(). It works because the repeated CallTo creates a new rule on which the applicator lives. The DoesNothing will leave the default applicator on the rule, and then it's used when the call is made.

Good point. I'm just back from holiday and my brained isn't fully turned back on yet :wink:

adamralph replied:
Do we we have any specs which verify that the sequence of actions described above work in the way described?

thomaslevesque replied:
> Do we we have any specs which verify that the sequence of actions described above work in the way described?

I just checked, we don't seem to have such specs.

blairconrad replied:
Are you talking about [these specs](https://github.com/blairconrad/FakeItEasy/blob/49ab38fd9fc06c4e5c00b2d2870e318e0178f9f6/tests/FakeItEasy.Specs/ConfigurationSpecs.cs#L495-L534)?

I added them when I fixed my fixed. I figured since @thomaslevesque had asked about them, I might as well add evidence that it worked the way I thought while I was at it.

thomaslevesque replied:
Ah, I missed those, thanks @blairconrad.

blairconrad replied:
Not at all. I'm glad everyone's keeping me on my toes!

blairconrad commented:
Okay. `DoesNothing` will (once again) not mix with `CallsBaseMethod` or `Throws` or `Returns` when invoked on the same rule.
I added a refactoring step to avoid duplicating the "default return value logic".
Once I'd added a method to set the applicator to the default behaviour, I noticed the property/method dichotomy that I'd introduced and made `Applicator` `UseApplicator` and removed the getter altogether, since it was only used by unit tests that were covered elsewhere (and so out they went).
It's more than I expected to do, and I'm happy to change any of it.

thomaslevesque commented:
The changes look good to me, and I like your `UseApplicator` refactoring :+1:

blairconrad commented:
Thanks, @thomaslevesque!

thomaslevesque commented:
@adamralph, any further comments?

adamralph commented:
I like the refactoring too, and the removal of the Gulliver's Travels tests.

Whilst we're here, I think it might be nice to lock in some of the sequential behaviour that we've mentioned in some specs, if it's not already done.

blairconrad commented:
I added a new commit because I'd accidentally used "it should do"-style naming instead of "it does"-style for some `Then` steps. I'd be inclined to eventually squash into the first commit where the specs were added.

adamralph commented:
After squashing the 4th commit into the 1st, I think this is good to go.

BTW - `git commit --fixup` and `git rebase --autosquash` are ideal for these scenarios. See https://robots.thoughtbot.com/autosquashing-git-commits

blairconrad commented:
Thanks for the `commit --fixup` and `rebase --autosquash` hint, @adamralph. I used them!
I'm not entirely sure it saved me any effort, but it was an interesting experiment. I'll try it a few more times to evaluate.

thomaslevesque commented:
Thanks @blairconrad!

blairconrad commented:
Thanks, y'all!
